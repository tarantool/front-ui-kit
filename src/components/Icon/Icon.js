// @flow
import * as React from 'react'
import { css, cx } from '@emotion/css';
import { rgba } from 'emotion-rgba';
import { colors, iconSize } from '../../variables';
import { SVGImage } from '../SVGImage';

const styles = {
  icon: css`
    flex-shrink: 0;
    vertical-align: middle;
    width: ${iconSize};
    height: ${iconSize};
  `,
  state: css`
    fill: red;

    &:hover {
      fill: greenyellow;
    }

    &.active {
      fill: blue;
    }
  `,
  stroke: css`
    stroke: red;

    &:hover {
      fill: greenyellow;
    }

    &.active {
      fill: blue;
    }
  `,
  clickable: css`
    cursor: pointer;
  `,
  active: css``,
  button: css`
    display: block;
    padding: 0;
    border: none;
    outline: none;
    background: transparent;

    &:focus::before {
      content: '';
      position: absolute;
      top: -2px;
      left: -2px;
      right: -2px;
      bottom: -2px;
      border: solid 1px ${rgba(colors.intentPrimary, 0.55)};
      border-radius: 3px;
    }
  `
};

type IconProps = {
  active?: boolean; // Выбраное состояние
  className?: string;
  glyph: SVGGlyph,
  hasState?: boolean, // Включение состояния: Normal, Hover, Active
  onClick?: (evt: MouseEvent) => void,
  onMouseEnter?: () => void,
  onMouseLeave?: () => void,
  stroke?: boolean // Задавать stroke вместо fill
};

export type GenericIconProps = {
  className?: string,
  onClick?: (e: MouseEvent) => void
};


export const Icon = React.memo<IconProps>(({
  active,
  className,
  glyph,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  hasState, // TODO: 'hasState' is defined but never used.
  onMouseLeave,
  onMouseEnter,
  onClick,
  stroke
}: IconProps) => (
  <SVGImage
    glyph={glyph}
    onClick={onClick}
    onMouseLeave={onMouseLeave}
    onMouseEnter={onMouseEnter}
    className={cx(
      styles.icon,
      {
        [styles.stroke]: stroke,
        [styles.clickable]: !!onClick,
        [styles.active]: active
      },
      className
    )}
  />
));
