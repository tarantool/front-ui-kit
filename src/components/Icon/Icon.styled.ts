import { css } from '@emotion/css';
import { rgba } from 'emotion-rgba';

import { colors, iconSize } from '../../variables';

export const styles = {
  icon: css`
    flex-shrink: 0;
    vertical-align: middle;
    width: ${iconSize};
    height: ${iconSize};
  `,
  state: css`
    fill: red;

    &:hover {
      fill: greenyellow;
    }

    &.active {
      fill: blue;
    }
  `,
  stroke: css`
    stroke: red;

    &:hover {
      fill: greenyellow;
    }

    &.active {
      fill: blue;
    }
  `,
  clickable: css`
    cursor: pointer;
  `,
  active: css``,
  button: css`
    display: block;
    padding: 0;
    border: none;
    outline: none;
    background: transparent;

    &:focus::before {
      content: '';
      position: absolute;
      top: -2px;
      left: -2px;
      right: -2px;
      bottom: -2px;
      border: solid 1px ${rgba(colors.intentPrimary, 0.55)};
      border-radius: 3px;
    }
  `,
};
