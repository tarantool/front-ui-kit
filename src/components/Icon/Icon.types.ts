import type { MouseEvent } from 'react';

export type GenericIconProps = {
  className?: string;
  onClick?: (event: MouseEvent) => void;
};

export type SVGGlyphTypes = {
  content: string;
  id: string;
  node: SVGSVGElement;
  viewBox: string;
};
