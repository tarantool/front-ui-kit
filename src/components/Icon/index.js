// @flow
import * as React from 'react'
import { css, cx } from 'emotion';
import { iconSize } from '../../variables';

const styles = {
  icon: css`
    vertical-align: middle;
    width: ${iconSize};
    height: ${iconSize};
  `,

  state: css`
    fill: red;

    &:hover {
      fill: greenyellow;
    }

    &.active {
      fill: blue;
    }
  `,

  stroke: css`
    stroke: red;

    &:hover {
      fill: greenyellow;
    }

    &.active {
      fill: blue;
    }
  `,

  clicked: css`
    cursor: pointer;
  `,

  active: css``
};

type Glyph = {
  content: string,
  id: string,
  node: any,
  viewBox: string,
};

interface IconProps {
  active?: boolean; // Выбраное состояние
  className?: string;
  glyph: Glyph,
  hasState?: boolean, // Включение состояния: Normal, Hover, Active
  onClick?: (evt: MouseEvent) => void,
  onMouseEnter?: () => void,
  onMouseLeave?: () => void,
  stroke?: boolean // Задавать stroke вместо fill
}

const Icon = ({
  active,
  className,
  glyph,
  hasState,
  onMouseLeave,
  onMouseEnter,
  onClick,
  stroke
}: IconProps) => {
  const isInteractive = onClick || onMouseEnter || onMouseLeave;

  const svg = (
    <svg
      viewBox={glyph.viewBox}
      className={cx(
        styles.icon,
        {
          [styles.state]: hasState,
          [styles.stroke]: stroke,
          [styles.clicked]: !!onClick,
          [styles.active]: active
        },
        className,
      )}
    >
      <use xlinkHref={`#${glyph.id}`}/>
    </svg>
  );

  return isInteractive
    ? (
      <button
        className={cx(
          styles.icon,
          {
            [styles.state]: hasState,
            [styles.stroke]: stroke,
            [styles.clicked]: !!onClick,
            [styles.active]: active
          },
          className,
        )}
        onClick={onClick}
        onMouseLeave={onMouseLeave}
        onMouseEnter={onMouseEnter}
        type='button'
      >
        {svg}
      </button>
    )
    : svg;
};

export default Icon;
