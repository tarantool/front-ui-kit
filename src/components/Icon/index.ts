export * from './Icon';
export * from './icons/IconArrow';
export * from './icons/IconCheckbox';
export * from './icons/IconChevron';
export * from './icons/IconSpinner';
