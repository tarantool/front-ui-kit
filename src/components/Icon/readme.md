Icon can be static or interactive

### Icon adding guide

Add new icon as a component in `src/components/Icon/icons` folder.
Use `IconChevron` as a reference. Basic concepts:

* Create new folder and name it `Icon[Name]`
* Place your code inside index.js
* Use https://jakearchibald.github.io/svgomg/ to optimize svg code
* Place svg beside js
* If icon have states (normal, hover, active), remove all `fill` attributes in svg and pass prop `hasState` to component

### Icons set

```
import { IconAttach } from './icons/IconAttach';
import { IconBox } from './icons/IconBox';
import { IconBoxNoData } from './icons/IconBoxNoData';
import { IconBucket } from './icons/IconBucket';
import { IconBurger } from './icons/IconBurger';
import { IconCancel } from './icons/IconCancel';
import { IconCheckbox } from './icons/IconCheckbox';
import { IconChevron } from './icons/IconChevron';
import { IconChip } from './icons/IconChip';
import { IconClose } from './icons/IconClose';
import { IconCluster } from './icons/IconCluster';
import { IconCrown } from './icons/IconCrown';
import { IconDownload } from './icons/IconDownload';
import { IconGear } from './icons/IconGear';
import { IconInfo } from './icons/IconInfo';
import { IconLink } from './icons/IconLink';
import { IconMore } from './icons/IconMore';
import { IconOk } from './icons/IconOk';
import { IconRadio } from './icons/IconRadio';
import { IconSearch } from './icons/IconSearch';
import { IconUser } from './icons/IconUser';
import { IconUsers } from './icons/IconUsers';

<div style={{ backgroundColor: 'darkgray' }}>
  <IconAttach />
  <IconBox />
  <IconBoxNoData />
  <IconBucket />
  <IconBurger />
  <IconCancel />
  <IconCheckbox />
  <IconChevron />
  <IconChevron direction='bottom' />
  <IconChip />
  <IconClose />
  <IconCluster />
  <IconCrown />
  <IconDownload />
  <IconGear />
  <IconInfo />
  <IconLink />
  <IconMore />
  <IconOk />
  <IconRadio />
  <IconSearch />
  <IconUser />
  <IconUsers />
</div>
```
