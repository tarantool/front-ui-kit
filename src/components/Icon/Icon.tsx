import React, { FC, MouseEvent } from 'react';
import { cx } from '@emotion/css';

import { SVGImage } from '../SVGImage';
import { styles } from './Icon.styled';

type IconProps = {
  active?: boolean; // Выбраное состояние
  className?: string;
  glyph?: SVGGlyphTypes;
  hasState?: boolean; // Включение состояния: Normal, Hover, Active
  onClick?: (evt: MouseEvent) => void;
  onMouseEnter?: () => void;
  onMouseLeave?: () => void;
  stroke?: boolean; // Задавать stroke вместо fill
};

export type GenericIconProps = {
  className?: string;
  onClick?: (e: MouseEvent) => void;
};

export type SVGGlyphTypes = {
  content: string;
  id: string;
  node: SVGSVGElement;
  viewBox: string;
};

export const Icon: FC<IconProps> = React.memo<IconProps>(
  ({
    active,
    className,
    glyph,
    // hasState, // TODO: 'hasState' is defined but never used.
    onMouseLeave,
    onMouseEnter,
    onClick,
    stroke,
  }: IconProps) => (
    <SVGImage
      glyph={glyph}
      onClick={onClick}
      onMouseLeave={onMouseLeave}
      onMouseEnter={onMouseEnter}
      className={cx(
        styles.icon,
        {
          [styles.stroke]: stroke,
          [styles.clickable]: !!onClick,
          [styles.active]: active,
        },
        className
      )}
    />
  )
);
