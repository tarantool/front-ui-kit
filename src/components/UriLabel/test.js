import React from 'react';
import renderer from 'react-test-renderer';
import { UriLabel } from './index';
import { IconGeoPin } from '../Icon';

it('Text renders correctly', () => {
  const tree = renderer
    .create(
      <>
        <UriLabel uri="jh74gvr3-ec3-datacenter-southern.awa132.com" title="UriLabel" />
        <UriLabel uri="jh74gvr3-ec3-datacenter-southern.awa132.com" title="UriLabel" weAreHere />
        <UriLabel uri="jh74gvr3-ec3-datacenter-southern.awa132.com" title="UriLabel" icon={IconGeoPin} />
      </>
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
