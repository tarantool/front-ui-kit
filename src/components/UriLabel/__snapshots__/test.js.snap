// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text renders correctly 1`] = `
Array [
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  width: 14px;
  height: 14px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  vertical-align: middle;
  width: 14px;
  height: 14px;
  width: 14px;
  height: 14px;
  fill: #5c606c;
  width: 14px;
  height: 14px;
  margin-right: 4px;
}

.emotion-2 {
  margin: 0;
  font-family: Inter,Arial,sans-serif;
  font-size: 12px;
  line-height: 20px;
  color: #000000;
  -webkit-font-smoothing: antialiased;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  color: rgba(0, 0, 0, 0.65);
}

<div
    className="emotion-0"
    title="UriLabel"
  >
    <svg
      className="emotion-1"
      viewBox="0 0 14 14"
    >
      <use
        xlinkHref="#link"
      />
    </svg>
    <span
      className="emotion-2"
    >
      jh74gvr3-ec3-datacenter-southern.awa132.com
    </span>
  </div>,
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  width: 16px;
  height: 16px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  vertical-align: middle;
  width: 14px;
  height: 14px;
  width: 16px;
  height: 16px;
  fill: #5c606c;
  fill: #ff272c;
  width: 14px;
  height: 14px;
  margin-right: 4px;
}

.emotion-2 {
  margin: 0;
  font-family: Inter,Arial,sans-serif;
  font-size: 12px;
  line-height: 20px;
  color: #000000;
  -webkit-font-smoothing: antialiased;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  color: rgba(0, 0, 0, 0.65);
}

<div
    className="emotion-0"
    title="UriLabel"
  >
    <svg
      className="emotion-1"
      viewBox="0 0 16 16"
    >
      <use
        xlinkHref="#icon-bold-arrow-right"
      />
    </svg>
    <span
      className="emotion-2"
    >
      jh74gvr3-ec3-datacenter-southern.awa132.com
    </span>
  </div>,
  .emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-1 {
  width: 14px;
  height: 14px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  vertical-align: middle;
  width: 14px;
  height: 14px;
  width: 14px;
  height: 14px;
  fill: #ff272c;
  width: 14px;
  height: 14px;
  margin-right: 4px;
}

.emotion-2 {
  margin: 0;
  font-family: Inter,Arial,sans-serif;
  font-size: 12px;
  line-height: 20px;
  color: #000000;
  -webkit-font-smoothing: antialiased;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  color: rgba(0, 0, 0, 0.65);
}

<div
    className="emotion-0"
    title="UriLabel"
  >
    <svg
      className="emotion-1"
      viewBox="0 0 14 14"
    >
      <use
        xlinkHref="#geo-pin"
      />
    </svg>
    <span
      className="emotion-2"
    >
      jh74gvr3-ec3-datacenter-southern.awa132.com
    </span>
  </div>,
]
`;
